File: stack/c/reverse_polish_notation.c
#include <stdio.h>
#include "reverse_polish_notation.h"

int parse_reverse_polish_notation(const char expression[])
{
  /* This parser doesn't check if the given expression is correct! */
  stack_t *s = create_stack();
  int a, b, res;

  for (unsigned int i = 0; expression[i] != '='; i++) {
    if (expression[i] >= '0' && expression[i] <= '9') {
      push(s, expression[i] - '0'); continue;
    }
    switch (expression[i]) {
      case '+':
        b = pop(s); a = pop(s);
        push(s, a + b);
      break;
      case '-':
        b = pop(s); a = pop(s);
        push(s, a - b);
      break;
      case '*':
        b = pop(s); a = pop(s);
        push(s, a * b);
      break;
      case '/':
        b = pop(s); a = pop(s);
        push(s, a / b);
      break;
    }
  }
  res = pop(s);
  remove_stack(s);
  return res;
}

void convert_to_reverse_polish_notation(const char expression[])
{
  stack_t *s = create_stack();
  for (unsigned int i = 0; expression[i] != '='; i++) {
    if (expression[i] >= '0' && expression[i] <= '9') {
      printf("%c", expression[i]); continue;
    }
    switch (expression[i]) {
      case '(':
        push(s, expression[i]);
      break;
      case ')':
        while (!empty_stack(s) && top(s) != '(')
          printf("%c", pop(s));
        pop(s);  /* Remove the opening bracket. */
      break;
      case '+':
      case '-':
      case '*':
      case '/':
        while (!empty_stack(s) && (top(s) == '*' || top(s) == '/'))
          printf("%c", pop(s));
        push(s, expression[i]);
      break;
    }
  }
  printf("%c", pop(s));
  remove_stack(s);
}

int main(void)
{
  // char expression[] = "352*+=";
  // printf("%i\n", parse_reverse_polish_notation(expression));

  char expression[] = "(5+3)*2/(2*5-3)=";
  convert_to_reverse_polish_notation(expression);

  return 0;
}

File: stack/python/reverse_polish_notation.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from typing import List, Dict, Callable

# types
Stack = List

def _add(s: Stack, a: int, b: int):
  s.append(a + b)

def _sub(s: Stack, a: int, b: int):
  s.append(a - b)

def _mul(s: Stack, a: int, b: int):
  s.append(a * b)

def _div(s: Stack, a: int, b: int):
  s.append(a / b)

def parse_reverse_polish_notation(expression: str) -> int:
  """Parses the reverse polish notation given in `expression` and
     evaluates it.
     NOTE: This parser doesn't check if the `expression` is correct!
  """
  OPERATIONS: Dict[str, Callable[Stack, int, int]] = {
    '+': _add,
    '-': _sub,
    '*': _mul,
    '/': _div
  }
  s: Stack[int] = []

  # Go through 0..n-1, where n is the last element.
  # We assumes that the last element is '='.
  for c in expression[:-1]:
    if '0' <= c <= '9':
      s.append(int(c))
      continue

    try:
      b = s.pop()
      a = s.pop()
      OPERATIONS[c](s, a, b)
    except KeyError:
      continue

  return s.pop()

def convert_to_reverse_polish_notation(expression: str) -> None:
  """Converts the given `expression` into the reverse polish notation."""
  s: Stack[str] = []

  # Go through 0..n-1, where n is the last element.
  # We assumes that the last element is '='.
  for c in expression[:-1]:
    if '0' <= c <= '9':
      print(c, end='')
      continue

    if c == '(':
      s.append(c)
    elif c == ')':
      while len(s) > 0 and s[-1] != '(':
        print('%c' % s.pop(), end='')
      s.pop()  # Remove the opening bracket
    elif c in ('+', '-', '*', '/'):
      while len(s) > 0 and (s[-1] == '*' or s[-1] == '/'):
        print('%c' % s.pop(), end='')
      s.append(c)

  print('%c' % s.pop(), end='')


if __name__ == '__main__':
  # print(parse_reverse_polish_notation('53+2*25*3-/='))
  convert_to_reverse_polish_notation('(5+3)*2/(2*5-3)=')
